canarytokens/webhook_formatting.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import annotations
+from typing import Union
+from enum import Enum
+import re
+from functools import partial
+from datetime import datetime
+
+from pydantic import HttpUrl, parse_obj_as
+
+from canarytokens import constants
+from canarytokens.channel import (
+    format_as_discord_canaryalert,
+    format_as_googlechat_canaryalert,
+    format_as_ms_teams_canaryalert,
+    format_as_slack_canaryalert,
+)
+from canarytokens.models import (
+    Memo,
+    TokenTypes,
+    TokenAlertDetails,
+    TokenExposedDetails,
+)
+
+
+WEBHOOK_TEST_URL = parse_obj_as(HttpUrl, "http://example.com/test/url/for/webhook")
+
+
+class WebhookType(Enum):
+    SLACK = "slack"
+    GOOGLE_CHAT = "google-chat"
+    DISCORD = "discord"
+    MS_TEAMS = "ms-teams"
+    GENERIC = "generic"
+
+
+def _match_url_start(url: str, match_str: str) -> bool:
+    return url.startswith(match_str)
+
+
+def _match_url_regex(url: str, match_regex: str) -> bool:
+    return bool(re.match(match_regex, url))
+
+
+_WEBHOOK_URL_MATCH_CRITERIA = {
+    WebhookType.SLACK: partial(
+        _match_url_start, match_str=constants.WEBHOOK_BASE_URL_SLACK
+    ),
+    WebhookType.GOOGLE_CHAT: partial(
+        _match_url_start, match_str=constants.WEBHOOK_BASE_URL_GOOGLE_CHAT
+    ),
+    WebhookType.DISCORD: partial(
+        _match_url_start, match_str=constants.WEBHOOK_BASE_URL_DISCORD
+    ),
+    WebhookType.MS_TEAMS: partial(
+        _match_url_regex, match_regex=constants.WEBHOOK_BASE_URL_REGEX_MS_TEAMS
+    ),
+}
+
+
+def get_webhook_type(url: str) -> WebhookType:
+    for webhook_type, match_func in _WEBHOOK_URL_MATCH_CRITERIA.items():
+        if match_func(url):
+            return webhook_type
+    else:
+        return WebhookType.GENERIC
+
+
+def format_details_for_webhook(
+    webhook_type: WebhookType, details: Union[TokenAlertDetails, TokenExposedDetails]
+):
+    if isinstance(details, TokenAlertDetails):
+        return _format_alert_details_for_webhook(webhook_type, details)
+    else:
+        return _format_exposed_details_for_webhook(webhook_type, details)
+
+
+def _format_alert_details_for_webhook(
+    webhook_type: WebhookType, details: TokenAlertDetails
+):
+    if webhook_type == WebhookType.SLACK:
```

Comment:
Could we use match and case here instead? Or are we not running python >= 3.10 here?

Reply:
Unfortunately we still run Python 3.9 on our tokens servers so we can't use match just yet.